fastchat/model/falcon_model.py

```diff
@@ -0,0 +1,138 @@
+import torch
+from typing import Iterable
+import gc
+from threading import Thread
+from transformers import TextIteratorStreamer, GenerationConfig
+
+import transformers
+
+transformers.logging.set_verbosity_error()
+
+
+def partial_stop(output, stop_str):
+    for i in range(0, min(len(output), len(stop_str))):
+        if stop_str.startswith(output[-i:]):
+            return True
+    return False
+
+
+@torch.inference_mode()
+def falcon_generate_stream(
+    model, tokenizer, params, device, context_len=2048, stream_interval=2
+):
+    prompt = params["prompt"]
+    len_prompt = len(prompt)
+    temperature = float(params.get("temperature", 1.0))
+    repetition_penalty = float(params.get("repetition_penalty", 1.0))
+    top_p = float(params.get("top_p", 1.0))
+    top_k = int(params.get("top_k", 50))  # -1 means disable
+    max_new_tokens = int(params.get("max_new_tokens", 256))
+    stop_str = params.get("stop", None)
+    echo = bool(params.get("echo", True))
+    stop_token_ids = params.get("stop_token_ids", None) or []
+    stop_token_ids.append(tokenizer.eos_token_id)
+
+    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)
+    input_ids = inputs["input_ids"]
+    attention_mask = inputs["attention_mask"]
+
+    max_src_len = context_len - max_new_tokens - 8
+
+    input_ids = input_ids[-max_src_len:]  # truncate from the left
+    attention_mask = attention_mask[-max_src_len:]  # truncate from the left
+
+    input_echo_len = len(input_ids)
+
+    decode_config = dict(skip_special_tokens=True, clean_up_tokenization_spaces=True)
+    streamer = TextIteratorStreamer(tokenizer, skip_prompt=True, **decode_config)
+
+    generation_config = GenerationConfig(
+        max_new_tokens=max_new_tokens,
+        do_sample=True,
+        temperature=temperature,
```

Comment:
When testing falcon7b-instruct, I received the following error.

```
2023-06-18 19:13:06 | ERROR | stderr |     raise ValueError(f"`temperature` has to be a strictly positive float, but is {temperature}")
2023-06-18 19:13:06 | ERROR | stderr | ValueError: `temperature` has to be a strictly positive float, but is 0.0
```

I temporarily hard-coded this to 1.0 and it worked.